### PR TITLE
TASK5: ready for review

### PR DESCRIPTION
**Pull request by Vadym-tst**
Signed-off-by: Vadym.Hutovskyi <vadym.hutovskyi@globallogic.com>

### CODE REVIEW

File: 05_timers/User space/user_space.c [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#define BUF_LENGTH				(99)
+#define TO_STRING(x)			(#x)
+#define CLOCK_ID_NUM			(8)
+
+static int clock_id[] = {CLOCK_REALTIME, CLOCK_REALTIME_COARSE,
+							CLOCK_MONOTONIC, CLOCK_MONOTONIC_COARSE,
+							CLOCK_MONOTONIC_RAW, CLOCK_BOOTTIME,
+							CLOCK_PROCESS_CPUTIME_ID, CLOCK_THREAD_CPUTIME_ID};
+
+static const char * const clock_id_name[] = {TO_STRING(CLOCK_REALTIME),
```
Comment (ziod) [inline 05_timers/User space/user_space.c:15]: clock_id and clock_id_name may be consumed in one struct.

File: 05_timers/User space/user_space.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		if (result != 0) {
+			printf("Error %d: can't get current time\n", result);
+			return 1;
+		}
+		printf("%s time is: %ld.%ld seconds\n", clock_id_name[i], time.tv_sec,
+																time.tv_nsec);
+		result = clock_getres(clock_id[i], &time);
+		if (result != 0) {
+			printf("Error %d: can't get current time", result);
+			return 1;
```
Comment (ziod) [inline 05_timers/User space/user_space.c:40]: 1; magic number

**Review by ziod — CHANGES_REQUESTED** [review]
pease do not use spaces in path.

File: 05_timers/Kernel space/kernel_space.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+#include <linux/err.h>
+#include <linux/string.h>
+#include <linux/sysfs.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/device.h>
+#include <linux/types.h>
+
+#define SYS_FOLDER_NAME					("Get_time_module")
+#define ABSOLUTE_TIME_ATTRIBUTE_NAME	Absolute_time
```
[the remark this answers is not in the document]
Comment (Vadym-tst) [reply 05_timers/Kernel space/kernel_space.c:14]: negative. That is not a string data type

File: 05_timers/Kernel/kernel_space.c [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+																previous_timestamp.tv_nsec);
+	ktime_get_real_ts64(&previous_timestamp);
+
+	return	strlen(buf);
+}
+
+static ssize_t relative_time(struct	kobject	*kobj, struct kobj_attribute *attr,	char *buf)
+{
+	static	struct timespec64 previous_timestamp;
+	struct	timespec64 current_timestamp, delta_time;
```
Comment (ziod) [inline 05_timers/Kernel/kernel_space.c:31]: Please do not define variables in coma separated way.

File: 05_timers/Kernel/kernel_space.c [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/device.h>
+#include <linux/types.h>
+
+#define	SYS_FOLDER_NAME					("Get_time_module")
+#define	ABSOLUTE_TIME_ATTRIBUTE_NAME	Absolute_time
+#define	RELATIVE_TIME_ATTRIBUTE_NAME	Relative_time
+
+static ssize_t absolute_time(struct	kobject	*kobj, struct kobj_attribute *attr,	char *buf)
```
Comment (ziod) [inline 05_timers/Kernel/kernel_space.c:17]: It is commn practice to name methods verb_subject.
show_absolufe_timestamp etc

File: 05_timers/Kernel/kernel_space.c [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+static int mod_init(void)
+{
+	int result	= 0;
+
+	time_module_kobject = kobject_create_and_add(SYS_FOLDER_NAME, kernel_kobj);
+	if	(!time_module_kobject) {
+		result = -ENOMEM;
+	} else	{
+		if (sysfs_create_file(time_module_kobject, &absolute_time_atr.attr))
+			result =	-ENOMEM;
```
Comment (ziod) [inline 05_timers/Kernel/kernel_space.c:56]: Unnecessary spaces

File: 05_timers/Kernel/kernel_space.c [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+		pr_err("error	no memory\n");
+		kobject_put(time_module_kobject);
+	}
+	return	result;
+}
+
+static void	mod_exit(void)
+{
+	sysfs_remove_file(time_module_kobject,	&absolute_time_atr.attr);
+	sysfs_remove_file(time_module_kobject,	&relative_time_atr.attr);
```
Comment (ziod) [inline 05_timers/Kernel/kernel_space.c:71]: Spaces

File: 05_timers/User/user_space.c [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+struct clock_id {
+	int clock_id_num;
+	const char * const clock_id_name;
+};
+
+int main(int argc, char *argv[])
+{
+	struct timespec time;
+	int result = 0;
+	struct clock_id clocks[CLOCK_ID_MAX] = {
```
Comment (ziod) [inline 05_timers/User/user_space.c:19]: Can be static const

File: 05_timers/User/user_space.c [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+				{CLOCK_MONOTONIC_RAW, TO_STRING(CLOCK_MONOTONIC_RAW)},
+				{CLOCK_BOOTTIME, TO_STRING(CLOCK_BOOTTIME)},
+				{CLOCK_PROCESS_CPUTIME_ID, TO_STRING(CLOCK_PROCESS_CPUTIME_ID)},
+				{CLOCK_THREAD_CPUTIME_ID, TO_STRING(CLOCK_THREAD_CPUTIME_ID)},
+			};
+
+	for (int i = 0; i < CLOCK_ID_MAX; i++) {
+		result = clock_gettime(clocks[i].clock_id_num, &time);
+		if (result != 0)
+			printf("Error %d: can't get current time\n", result);
```
Comment (ziod) [inline 05_timers/User/user_space.c:33]: It is better to keep one empty line after to increase readability

**Review by ziod — CHANGES_REQUESTED** [review]
Uninformatve commit message

File: 05_timers/Kernel/kernel_space.c [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	ktime_get_real_ts64(&previous_timestamp);
+
+	return	strlen(buf);
+}
+
+static ssize_t relative_time(struct	kobject	*kobj, struct kobj_attribute *attr,	char *buf)
+{
+	static	struct timespec64 previous_timestamp;
+	struct	timespec64 current_timestamp;
+    struct	timespec64 delta_time;
```
Comment (ziod) [inline 05_timers/Kernel/kernel_space.c:32]: please align

**Review by ziod — CHANGES_REQUESTED** [review]
please resolve all the comments

**Review by ziod — APPROVED** [review]

**Pull request merged**